who_is_active.sql

```diff
@@ -2740,7 +2741,15 @@ BEGIN;
 													END +
 												N'')''
 											WHEN y.wait_type = N''CXPACKET'' THEN
-												N'':'' + SUBSTRING(y.resource_description, CHARINDEX(N''nodeId'', y.resource_description) + 7, 4)
+												N'':'' + SUBSTRING(y.resource_description, 

```

Comment:
Do we need both CXPACKET and CXCONSUMER sections here? They look identical? How about IN ('CXPACKET', 'CXCONSUMER') instead?

Reply:
I don't think this works properly. When I run the proc with a parallel workload and @get_task_info = 2, I'm getting "Invalid length parameter passed to the LEFT or SUBSTRING function."

I debugged a bit and I think the parsing bit isn't correct:

```
declare @resource_description varchar(100) = 'exchangeEvent id=Pipe1feb489cb80 WaitType=e_waitPipeNewRow nodeId=0'

select
SUBSTRING(@resource_description, 
																   CHARINDEX(N'nodeId', @resource_description) + 7, 
																   CHARINDEX(N' ', @resource_description, CHARINDEX(N'nodeId', @resource_description) + 7)
																   - 7 - CHARINDEX(N'nodeId', @resource_description))
```

Can you take a look?

Reply:
> Do we need both CXPACKET and CXCONSUMER sections here? They look identical? How about IN ('CXPACKET', 'CXCONSUMER') instead?

We could yeah, I had suggested that somewhere (maybe just via email, I can't recall at this point). But we should probably also include `CXSYNC_PORT` and `CXSYNC_CONSUMER` to be Azure/MI friendly.

Reply:
> I don't think this works properly. When I run the proc with a parallel workload and @get_task_info = 2, I'm getting "Invalid length parameter passed to the LEFT or SUBSTRING function."
> 
> I debugged a bit and I think the parsing bit isn't correct:
> 
> ```
> declare @resource_description varchar(100) = 'exchangeEvent id=Pipe1feb489cb80 WaitType=e_waitPipeNewRow nodeId=0'
> 
> select
> SUBSTRING(@resource_description, 
> 																   CHARINDEX(N'nodeId', @resource_description) + 7, 
> 																   CHARINDEX(N' ', @resource_description, CHARINDEX(N'nodeId', @resource_description) + 7)
> 																   - 7 - CHARINDEX(N'nodeId', @resource_description))
> ```
> 
> Can you take a look?
> 
> BTW, the tabbing is maybe a touch more intense than necessary.

Read back through our email thread, and I need to look at this more closely to figure out why I had to make this change. I'd guess because `CXCONSUMER` entries are formatted differently.

Reply:
Maybe a dumb question -- what version are you testing on? None of the parallel detail I have in `dm_os_waiting_tasks` ends the way yours does:

![image](https://user-images.githubusercontent.com/2136037/126700482-b3e8896c-ab94-446a-b22f-b5ee87823238.png)